### ISSUE TITLE
Deferred deeplinking not working

### ISSUE THREAD

**Issue by hearbeathorror** [adjust/android_sdk#360]
Hi,

I have implemented deferred deeplinking in an android application, but then the method does not seem to work since after june.

I have the following method implemented and it returns true default. However, any ad from facebook does not seem to trigger this link, and the method is not called.

AdjustConfig config = new AdjustConfig(this, appToken, environment);

`
    config.setOnDeeplinkResponseListener(new OnDeeplinkResponseListener() {
        @Override
        public boolean launchReceivedDeeplink(Uri deeplink) {
            return true;
        }
    });
`

please guide as to what might be missing or wrong? Or if there is any update in the sdk that we need to consider?

**Comment by uerceg** [comment]
Hi @hearbeathorror 

That part of the code is unchanged ever since introduced to the SDK, so I don't think that update would change anything (even though update to latest SDK version is always recommended, but at the end of the day it's up to you).

What exactly doesn't work? Method doesn't get triggered at all, or it does, but activity which handles given scheme is not being launched or something else?

Our SDK is pretty dummy when it comes deferred deep linking scenario. After an install has been tracked, in communication with our backend we will at some point get information from them whether attribution assigned to the user came in connection with some deep link attached (either to Adjust tracker URL with `deep_link` parameter or in your case, if FB lets us know that there was deep link also involved in click that drove that app install). After SDK gets deep link info, we do following:

```
if (deep link callback implemented) {
    launch deferred deep link callback with deep link;
    if (callback return value == true) {
        launch activity which handles scheme from deep link;
    }
} else {
    launch activity which handles scheme from deep link;
}
```

There's nothing more than that. So if callback is implemented and SDK is not triggering it, it's probably because deferred deep link info never came to SDK from backend. And why it didn't come, well in there scenarios can probably be numerous. This is something which most probably account management team would be able to help you in debugging by checking your app's traffic and logs.

But simple way to test if everything SDK wise works well:

- Create dummy tracker in your dashboard and make URL that looks something like: https://app.adjust.com/abc123?deep_link=your-scheme%3a%2f%2fwhatever where `abc123` is your dummy tracker token.
- Uninstall your app from test device.
- Forget the device with test console.
- You can put this link into some random website (I tend to use neocities.org to create dummy web page just to put these links in there).
- Open that web page from your test device and click onto this URL.
- If redirected to the store, ignore that.
- Set SDK to sandbox environment and log level to verbose.
- Install debug version of your app on test device with above mentioned SDK setup.
- Check the logs and check if your callback gets triggered.

In logs you should see that when providing final attribution, our backend also sent `deeplink` parameter to SDK which we took and did the logic explained in pseudo code above.

If this works well, no issues in your app and SDK integration. But what might be wrong, like said, it would then the best if you can check that with our support team.

Lemme know if you have any other questions.

Cheers

**Comment by rrobbie1** [comment]
Hi @uerceg 
As above, 
Is there a test method in the production environment?

**Comment by uerceg** [comment]
Hi @rrobbie1 

Well, everything pretty much works in exactly same way in production as well, just that SDK doesn't print any logs in production environment. Maybe one of the ways for you to check this can be to put tools like Charles proxy in between your device and your local network and simply check the content of the response you get when SDK asks for attribution.

I'll close this issue since I assume it's working for you having in mind issue you opened at RN repo. Feel free to reopen it in case issue still persists.

Cheers!

**Comment by boylenssen** [comment]
I've tried everything, but unable to get the deeplink after the app is installed.

When the app is already installed, everything works perfectly! But when I get redirected to the Play Store, install the app (via `adb install`) and then press open, there is no data in the intent.

I've also added `setOnDeeplinkResponseListener` to see if that would be triggered => it is not called.

I've tried SDK version 4.20 and 4.21.

Let me know what I can do to debug this further

**Comment by uerceg** [comment]
@boylenssen Did you forget your test device between uninstalling an app from it and clicking on tracker URL?

**Comment by boylenssen** [comment]
"forget" it in the adjust testing console? Yes, I did.
I removed the app, 'forget device', clicked the track url and installed the app via ADB

**Comment by uerceg** [comment]
Yes, that's what I meant. Okay. How does the tracker URL look like?

**Comment by boylenssen** [comment]
https://app.adjust.com/<id>?deep_link=<our_scheme>%3A%2F%2F<our_action_name>%2F<some_id>%2F<some_other_id

**Comment by uerceg** [comment]
Did you accidentally remove tracker token from URL or is this how it's really formatted? Usually it should look something like https://app.adjust.com/{token}? everything_else

In case you do have the token in URL, do you have fingerprinting for clicks in attribution settings enabled for it? If not, you either need to enable that or explicitly add your test device identifier to URL (&gps_adid=[your-device-google-advertising-id]).

**Comment by boylenssen** [comment]
Yes, sorry, I accidentally removed it! 
https://app.adjust.com/<tracker_token>?deep_link=<our_scheme>%3A%2F%2F<our_action_name>%2F<some_id>%2F<some_other_id>

I'm going to give the gps_adid a try!

**Comment by uerceg** [comment]
Sounds good, let us know if it worked.

**Comment by boylenssen** [comment]
Wow! Adding the adid to the URL worked!

What's next, is this something just for testing environment? Should I expect it to just work for "real world" users?

**Comment by boylenssen** [comment]
we need to get the deeplink any time, not just one time per adid, is that possible?

**Comment by uerceg** [comment]
Testing you just did is environment agnostic. It should not matter if you are doing this with debug build of an app from Android Studio or with real app published in the Play Store. So if this worked for you, it should work for users of your published app as well.

If fingerprinting for clicks was enabled, in theory your test should have worked even without appending gps_adid parameter to tracker URL. In case it didn't, that sounds weird, since it shouldn't be the case. If it still fails to do the fingerprinting, sounds like a question for your dedicated account manager (if you have one) or support@adjust.com (if you don't) to maybe dig a bit deeper to our logs and figure out what might be misconfigured so that this didn't work.

Deferred deep link (which is the scenario you are testing) will be delivered to the app only after your user installed it for the first time (since this is the install in Adjust sense of things). This is how things currently are. In case you would like to see this happening after reinstalls as well, we would need to update logic on our end, but that's already product decision. In case you have interest for such a feature, feel free to request it (same address as I mentioned above for your question) and product team will surely take a look at it and evaluate it.

**Comment by boylenssen** [comment]
I've send a link to this issue here to our dedicated account manager. Will wait on her response...

Thanks a lot so far!

**Comment by wangjianpeng** [comment]
hi, my deferred deep link only works in sandbox environment , but does not work in product environment, do you have the same problem ?

**Comment by uerceg** [comment]
Hi @wangjianpeng 

That should not be the case since deferred deep linking (and deep linking in general) has absolutely no dependency on which environment you're running SDK in. Environment you're setting in SDK is only used to pass information about that in each package we send to backend so that it knows should that traffic be counted as sandbox or production one (and we're shutting logging OFF in production).

But deferred deep linking should work without any problems in both environments. Did you make sure to follow steps from [this reply](https://github.com/adjust/android_sdk/issues/360#issuecomment-439354884) **every time** you're about to test deferred deep linking scenario?

Cheers

**Comment by luisafcastano** [comment]
@uerceg, regarding your comment

> Deferred deep link (which is the scenario you are testing) will be delivered to the app only after your user installed it for the first time (since this is the install in Adjust sense of things). This is how things currently are. In case you would like to see this happening after reinstalls as well, we would need to update logic on our end, but that's already product decision. In case you have interest for such a feature, feel free to request it (same address as I mentioned above for your question) and product team will surely take a look at it and evaluate it.

I want to double check: **This has changed, right?**

According to the [new docs](https://help.adjust.com/en/article/deferred-deep-linking-on-reinstall):
> You can use deferred deep links to retarget users who've uninstalled your app. The deferred deep link sends users to a specific place in your app after first routing them through the relevant app store to reinstall the app.

Just want to confirm it is the case. Thanks!

**Comment by jishnupd13** [comment]
@uerceg setOnAttributionChangedListener is not triggered in my case. Can you help me?

**Comment by simonmitchell** [comment]
Hi @onursumakoglu did you ever get this working? I'm having the same issue in v5 of the SDK.

**Comment by onursumakoglu** [comment]
hi @simonmitchell , actually situation is not changed. when I try it with advertising id it was working. I completed my tests like this way. in production I didn't faced any issue with url which is not have adv id
